### PR TITLE
ENH: add bdschur, which calls Slycot mb03rd.  Change modal_form to us…

### LINKED ISSUE #318: modal canonical form does not work with repeated eigenvalues
The `modal_form()` function appears to assume that all eigenvalues are distinct.  The following example gives strange results:
```
import control as ct
sys_repeat = ct.StateSpace([[-1, 1], [0, -1]], [[0], [1]], [[1, 0]], [[0]])
sys_modal, T = ct.modal_form(sys_repeat)
print(sys_modal)
```
which yields
```
A = [[-1.00000000e+00  0.00000000e+00]
 [ 2.22044605e-16 -1.00000000e+00]]

B = [[4.50359963e+15]
 [4.50359963e+15]]

C = [[-1.  1.]]

D = [[0.]]
```
The problem is that the transformation `T` as computed in `modal_form` is singular:
```
[[-1.00000000e+00  1.00000000e+00]
 [ 2.22044605e-16  0.00000000e+00]]
```

### PR DESCRIPTION
**Pull request by roryyorke**
…e bdschur

The bdschur interface is modelled after the Matlab function of the
same name; it can also optionally sort the modal blocks.

Added tests for bdschur, and modified tests for modal_form.

Closes #318

### CODE REVIEW

File: control/canonical.py [hunk 1]
```diff
@@ -149,97 +153,294 @@ def observable_form(xsys):
...
+_PMAX_SEARCH_TOL = 1.001
+
+
+def _bdschur_defective(blksizes, eigvals):
+    """Check  for defective modal decomposition
+
+    Parameters
+    ----------
+    blksizes: size of Schur blocks
+    eigvals: eigenvalues
```
Comment (bnavigator) [inline control/canonical.py:211]: It's a private function, but use numpydoc anyway ( `blksizes : array_like of ...`)?

File: control/canonical.py [hunk 2]
```diff
@@ -149,97 +153,294 @@ def observable_form(xsys):
...
+    """Block-diagonal Schur decomposition search up to condmax
+
+    Iterates mb03rd with different pmax values until:
+      - result is non-defective;
+      - or condition number of similarity transform is unchanging despite large pmax;
+      - or condition number of similarity transform is close to condmax.
+
+    Parameters
+    ----------
+    aschur: (n, n) array
```
Comment (bnavigator) [inline control/canonical.py:248]:
```suggestion
    aschur: (n, n) array_like
```

File: control/canonical.py [hunk 3]
```diff
@@ -149,97 +153,294 @@ def observable_form(xsys):
...
+    sort: False (default)
+        If true, Schur blocks will be sorted.  See `bdschur` for sort order.
 
     Returns
     -------
     zsys : StateSpace object
-        System in transformed coordinates, with state 'z'
-
+        System in modal canonical form, with state `z`
+    T : matrix
```
Comment (bnavigator) [inline control/canonical.py:433]: matrix?

File: control/tests/canonical_test.py [hunk 4]
```diff
@@ -258,3 +190,250 @@ def test_similarity(self):
...
+
+@pytest.mark.parametrize(
+    "eigvals, condmax, blksizes",
+    [
+        ([-1,-2,-3,-4,-5], None, [1,1,1,1,1]),
+        ([-1,-2,-3,-4,-5], 1.01, [5]),
+        ([-1,-1,-2,-2,-2], None, [2,3]),
+        ([-1+1j,-1-1j,-2+2j,-2-2j,-2], None, [2,2,1]),
+    ])
+def test_bdschur_ref(eigvals, condmax, blksizes):
```
Comment (bnavigator) [inline control/tests/canonical_test.py:253]: I guess most of these tests must be marked `@slycotonly`. see Travis CI fails

**Review by bnavigator — COMMENTED** [review]
Nice! Some comments about the docstrings and test suite. Can't say much about the algorithms themselves.

**Comment by roryyorke** [comment]
Thanks @bnavigator !  I've added the `@slycotonly` decorator, and updated the docs.  I did some minor mods to the other docstrings in `canonical.py` too, and added `bdschur` and `modal_form` to control.rst.

**Comment by murrayrm** [comment]
Rebased and changed `_convertToStateSpace` to `_convert_to_statespace`.  Will merge as soon as Travis finishes up.

**Pull request merged**